com.unity.ml-agents/Runtime/Actuators/ActionSegment.cs

```diff
@@ -0,0 +1,113 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Unity.MLAgents.Actuators
+{
+    public readonly struct ActionSegment<T> : IEnumerable<T>
+    {
+        public readonly int Offset;
+        public readonly int Length;
+
+        public static ActionSegment<T> Empty = new ActionSegment<T>(System.Array.Empty<T>(), 0, 0);
+
+        static void CheckParameters(T[] actionArray, int offset, int length)
+        {
+#if DEBUG
+            if (offset + length > actionArray.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset), $"Arguments offset: {offset} and length: {length} " +
+                    $"are out of bounds of actionArray: {actionArray.Length}.");
+            }
+#endif
+        }
+
+        public ActionSegment(T[] actionArray, int offset, int length)
+        {
+            CheckParameters(actionArray, offset, length);
+            Array = actionArray;
+            Offset = offset;
+            Length = length;
+        }
+
+        public T[] Array { get; }
```

Comment:
Is this needed anywhere? Seems like it breaks the encapsulation of the segmenting.

Reply:
I used it for copying arrays for[ backward compatibility](https://github.com/Unity-Technologies/ml-agents/blob/17540f61f2f366bea66eb35f7b82466236bd152d/com.unity.ml-agents/Runtime/Agent.cs#L1191) in `OnActionReceived` and `Heuristic`, which isn't in this PR.